portal/src/graphql/portal/SingleSignOnConfigurationScreen.tsx

```diff
@@ -0,0 +1,375 @@
+): OAuthSSOProviderConfigState {
+  const providerConfig = providerByType[providerType] ?? {
+    type: providerType,
+  };
+  // FIXME: make alias required
+  const alias = providerConfig.alias ?? providerType;
+  return {
+    ...providerConfig,
+    enabled: !!providerByType[providerType],
+    clientSecret: clientSecretByAlias[alias],
+  };
+}
+
+function extractOAuthSecret(secretConfig: PortalAPISecretConfig) {
+  for (const secret of secretConfig.secrets) {
+    if (secret.key === "sso.oauth.client") {
+      return secret;
+    }
+  }
+  return null;
+}
+
+function constructScreenState(
+  providers: OAuthSSOProviderConfig[],
+  oauthCredentials: OAuthClientCredentialItem[]
+) {
+  const providerByType = providers.reduce(
+    (
+      map: Partial<Record<OAuthSSOProviderType, OAuthSSOProviderConfig>>,
+      provider
+    ) => {
+      map[provider.type] = provider;
+      return map;
+    },
+    {}
+  );
+
+  const clientSecretByAlias = oauthCredentials.reduce(
+    (
+      map: Partial<Record<string, string>>,
+      credential: OAuthClientCredentialItem
+    ) => {
+      if (credential.alias !== "") {
+        map[credential.alias] = credential.client_secret;
+      }
+      return map;
+    },
+    {}
+  );
+
+  const screenState = oauthSSOProviderTypes.reduce(
+    (
+      map: Partial<
+        { [key in OAuthSSOProviderType]: OAuthSSOProviderConfigState }
+      >,
+      providerType
+    ) => {
+      map[providerType] = getProviderConfigStateFromProviders(
+        providerType,
+        providerByType,
+        clientSecretByAlias
+      );
+      return map;
+    },
+    {}
+  );
+
+  return screenState as SingleSignOnScreenState;
+}
+
+function constructProvidersFromState(state: SingleSignOnScreenState) {
+  const providers: OAuthSSOProviderConfig[] = [];
+  for (const providerType of oauthSSOProviderTypes) {
+    const providerConfigState = state[providerType];
+    if (!providerConfigState.enabled) {
+      continue;
+    }
+    providers.push({
+      type: providerConfigState.type,
+      alias: ensureNonEmptyString(providerConfigState.alias),
+      client_id: ensureNonEmptyString(providerConfigState.client_id),
+      tenant: ensureNonEmptyString(providerConfigState.tenant),
+      key_id: ensureNonEmptyString(providerConfigState.key_id),
+      team_id: ensureNonEmptyString(providerConfigState.team_id),
+    });
+  }
+
+  return providers;
+}
+
+function constructOAuthCredentialsFromState(state: SingleSignOnScreenState) {
+  const credentials: OAuthClientCredentialItem[] = [];
+
+  oauthSSOProviderTypes.forEach((provider) => {
+    const { enabled, clientSecret, alias } = state[provider];
+    if (enabled && clientSecret != null && clientSecret.trim() !== "") {
+      credentials.push({
+        // FIXME: make alias required
+        alias: alias ?? provider,
+        client_secret: clientSecret,
+      });
+    }
+  });
+
+  return credentials;
+}
+
+function constructViolationMap(
```

Comment:
The reason for preprocessing the violations is because we need to figure out which widget the violation belongs to. The pointer has the form `identity/oauth/providers/{index}`. The index refers to index in the array provided to update the config, and the widget does not know it's own `index`. `index` could be different in each request. (currently each provider type has 1 widget, fixed order)

I guess if we update UI design to follow the data structure this is possible (list of widget = list of providers in order, with add button to add widget instead of toggle). In this case widget index is given by the `.map()`. Currently this is the only solution I can come up with

Also there is general kind violation without position, not all error points to one specific field, so I added unhandled violation list. @kiootic

Reply:
Understood, let's open an issue to follow up on error presentation later when needed.

Reply:
I dont know the current arch but one approach I had before is to treat error as part of the entry model, so `UsernameInput` should contains `input: String` and `error: T?` (don't know typescript's error type, `T` suffices here). And then there would a bare `Input<T, E>` thing which simplify the boilerplate. And then we can go `input.validate()`, `input.dryValidate()`, for validation, and `input.validity` for checking.

Making each error type for each input unique (so we would have `UsernameInputError`) makes identifying source of error as simple as type checking.

(Actually I used to do validation with enum but the idea is the same)

Reply:
@roxk 

We created a interface called Violation, with a `location` property that should points to a field in config file (depends on kind of error), so the UI must match with data structure of config file such that each input field can select the related violations

On UI, we are using `TextField` component from FluentUI, and it takes `errorMessage: string | undefined` as props. 

As it takes error message as string, I didn't define error class for each field. We decided to define a selector, to select `violation` for each field, format list of `violation` to localized error message. I think `Violation` is similar to the error which is unique to each input field, but we check the location property instead of type.

@kiootic points out that there are two selections instead of one, first selection select violations for each widget, then second selection in widget selects violation for each field, and first selection should not exist.

As I explained above, the problem is widget does not have enough info to select violation using the location, as the UI and config data structure does no match up